### ISSUE TITLE
Tracking issue for dyn upcasting coercion

### ISSUE THREAD

**Issue by alexreg** [rust-lang/rust#65991]
This is a tracking issue for [RFC3324](https://github.com/rust-lang/rfcs/pull/3324). Corresponding [MCP](https://github.com/rust-lang/lang-team/issues/98) is here.
The feature gate for the issue is `#![feature(trait_upcasting)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement the experimental feature behind a feature gate.
- [x] Define the semantics explicitly.
- [x] Gathering design docs into initiative repo: https://github.com/rust-lang/dyn-upcasting-coercion-initiative
- [x] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide]) https://github.com/rust-lang/reference/pull/1259
- [x] Gather data about memory usage? https://github.com/rust-lang/rust/issues/112355
- [x] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide]) ~~https://github.com/rust-lang/rust/pull/101718~~, https://github.com/rust-lang/rust/pull/118133
- [x] Fix soundness problems.
  - https://github.com/rust-lang/rust/pull/120248
- [ ] Restabilize.

### Previous discussions
* RFC401: https://github.com/rust-lang/rfcs/pull/401
* RFC982: https://github.com/rust-lang/rfcs/pull/982
* RFC-issue2765: https://github.com/rust-lang/rfcs/issues/2765
* MCP: https://github.com/rust-lang/lang-team/issues/98 (Current lang-team MCP)
* Initiative: [Dyn upcasting coercion initiative](https://github.com/rust-lang/dyn-upcasting-coercion-initiative)
* RFC: https://github.com/rust-lang/rfcs/pull/3324

### Unresolved Questions
* ~~Currently `CoerceUnsized` trait cannot express this case: a smart pointers that wrap a raw pointer and don't guarantee via a custom invariant that it is valid. Maybe a separate `CoerceUnsizedUnsafe` trait is needed. (see https://github.com/rust-lang/rust/pull/88010#issuecomment-898893209).~~
* Should we make upcasting opt-in in some form to limit vtable size by default? The current inclination of the lang-team is "no", but it would be useful to gather data on how much supporting upcasting contributors to overall binary size.
* ~~Before stabilizing it we should check that libs-api is ok with upcasting for all dyn-allowed traits in the library, since those we can't change.~~ (addressed by https://github.com/rust-lang/rust/issues/65991#issuecomment-1369970129)
* Should we add an *opt-out* mechanism, and extend library stabilization checklist with "do we want to opt-out for now"?

### Implementation history
* Initial attempt: #60900
* Second attempt:
  * [X] Preparation refactoring: #86291
  * [X] Unify vtable building: #86475
  * [x] Update vtable format: #86461
  * [x] Feature gate implementation: #86264
  * [x] Pointer conversion adjustments: #87515
  * [x] Support multiple occurrence of a target trait with different substsdefs: #88135
  * [x] ~~Unsafety checking for raw dyn pointers: #88010 (not prefered approach, likely will need to do nothing here, but need a design meeting discussion with lang-team)~~ Solution: https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/meeting.202022-09-02.20planning/near/296826711, https://github.com/rust-lang/rust/issues/101336
  * [x] Lint for easing migration: https://github.com/rust-lang/rust/issues/89460
  * [x] Reference changes
    * [x] ~~https://github.com/rust-lang/reference/pull/1259~~
    * [ ] https://github.com/rust-lang/reference/pull/1622
  * [x] Mark this feature no longer incomplete and stablize: https://github.com/rust-lang/rust/pull/101718
  * [x] Make trait upcasting work with supertraits with fewer associated types: #114036

**Comment by nikomatsakis** [comment]
This is currently blocked on resolving the safety conditions for upcasting. The following updated document describes my current thinking, looking for feedback:

https://rust-lang.github.io/dyn-upcasting-coercion-initiative/design-discussions/upcast-safety-2.html

**Comment by dimpolo** [comment]
I was playing around with this and discovered that the feature gate (and its incomplete_features warning) can be bypassed using `#![feature(unsize)]` instead.

This will upcast a `&dyn Bar` where `Foo` is a supertrait of `Bar`.
```rust
fn upcast<Dyn: ?Sized + Unsize<dyn Foo>>(bar: &Dyn) -> &dyn Foo { bar }
```

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0719de3ae8996a57fc01f2afb0354001)

Would this mean the `unsize` feature should be incomplete as well?

**Comment by danielhenrymantilla** [comment]
Came to comment the same observation that @dimpolo has made: I think the impl of `Unsize<dyn '_ + Super>` for `dyn Sub` ought, itself, to be feature-gated by `trait_upcasting`.

**Comment by Mark-Simulacrum** [comment]
cc https://github.com/rust-lang/rust/issues/89460, which is tracking a future-compat lint that seems related to this feature.

**Comment by pnkfelix** [comment]
@joshtriplett can you confirm that when you added the S-waiting-on-team label, that was a reference to T-lang, and not T-compiler?

**Comment by pnkfelix** [comment]
@rustbot label: I-lang-nominated 

I am seeking an answer to the question, requoted here:
> @joshtriplett can you confirm that when you added the S-waiting-on-team label, that was a reference to T-lang, and not T-compiler?

**Comment by nikomatsakis** [comment]
This is blocked on lang decisions, and @crlf0710 has a stabilization PR ready to go

**Comment by apiraino** [comment]
Signaling the [above comment](https://github.com/rust-lang/rust/issues/65991#issuecomment-1238528625) with a `S-blocked` label

@rustbot label s-blocked

**Comment by dtolnay** [comment]
@rustbot label -S-blocked

Niko's comment was _"this is blocked on lang decisions"_. This issue was already labelled https://github.com/rust-lang/rust/labels/S-waiting-on-team https://github.com/rust-lang/rust/labels/T-lang which captures this &mdash; _"awaiting decision from the relevant subteam"_. https://github.com/rust-lang/rust/labels/S-blocked means something different &mdash; _"blocked on something else such as an RFC or other implementation work"_.

**Comment by apiraino** [comment]
@dtolnay it was just a convenience when I scan every week issues waiting on team to find the oldest ones to mention [for the T-compiler meeting](https://hackmd.io/P5_PrNfwR6Sp89yQ_S3-jg#Oldest-PRs-waiting-for-review).

but it's fine, I won't bikeshed labels meaning :)

**Comment by CAD97** [comment]
Something to consider before stabilization: while `trait A: B` is upcastable, should `trait A where Self: B` be?

I want to propose that perhaps `where Self: B` should *not* be considered an upcastable supertrait bound. That these two forms are equivalent is AIUI only apparent in one other location of the language -- that `where Self:` bounds are elaborated (on `trait` items), but [all other `where` bounds are not](https://github.com/rust-lang/rust/issues/103387).

TBQH I'm unsure one way or the other, but this should be explicitly decided, and I don't think it has been.

**Comment by jeffs** [comment]
> Something to consider before stabilization: while `trait A: B` is upcastable, should `trait A where Self: B` be?

Isn't the former syntactic sugar for the latter?  From [The Rust Reference](https://doc.rust-lang.org/reference/items/traits.html#supertraits):

> The following is an example of declaring Shape to be a supertrait of Circle.
>
> ```rust
> trait Shape { fn area(&self) -> f64; }
> trait Circle : Shape { fn radius(&self) -> f64; }
> ```
>
> And the following is the same example, except using [where clauses](https://doc.rust-lang.org/reference/items/generics.html#where-clauses).
> 
> ```rust
> trait Shape { fn area(&self) -> f64; }
> trait Circle where Self: Shape { fn radius(&self) -> f64; }
> ```

Any semantic difference between these syntaxes would be, at the least, surprising.

**Comment by CAD97** [comment]
> Isn't the former syntactic sugar for the latter?

Yes, that is currently the case. `where Self: Bound` is also special because it gets elaborated as a supertrait bound, unlike say `where Self::Assoc: Bound` which doesn't get elaborated.

I'm not convinced either way that `where Self: Bound` should or shouldn't offer upcasting; I think either are a valid choice, I just feel that the choice should be deliberately made rather than an artifact of the two forms currently being unified.

**Comment by nikomatsakis** [comment]
The RFC has been merged. I updated this issue with unresolved questions:

* Should we make upcasting opt-in in some form to limit vtable size by default? The current inclination of the lang-team is "no", but it would be useful to gather data on how much supporting upcasting contributors to overall binary size.
* Before stabilizing it we should check that libs-api is ok with upcasting for all dyn-allowed traits in the library, since those we can't change.
* Should we add an *opt-out* mechanism, and extend library stabilization checklist with "do we want to opt-out for now"?

To that end, it would be useful 

* if someone wanted to try and gather data about size overhead as a result of this change; you could compare the vtable behavior from before/after the improvements landed, or otherwise examine output.
* to prepare a list of libcore traits that are affected

**Comment by nikomatsakis** [comment]
@scottmcm raised the point in a previous meeting that we should review the set of traits that are affected.  @crlf0710 was awesome enough to do an analysis, [results available here](https://docs.google.com/spreadsheets/d/1zIKz5uWP2nLXdQmY4vHJhNXydbLK22OpolqbYYulesg/edit?usp=sharing). There will be 11 stable API traits that are now dyn upcastable, but only one of them has multiple supertraits and hence would be committed to the larger vtable:

* Error <-- has two supertraits, so increases vtable size
* BufRead
* BorrowMut
* PartialOrd
* IndexMut
* Fn
* FnMut
* DerefMut
* DoubleSidedIterator
* ExactSizeIterator
* FusedIterator

In addition there are 6 unstable traits that become upcastable (purple background), but none of them have multiple supertraits:

* std::str::pattern::ReverseSearcher
* std::str::pattern::DoubleEndedSearcher
* std::ops::OneSidedRange
* std::iter::traits::marker::TrustedLen
* std::simd::SimdPartialOrd
* std::simd::SimdOrd

21 traits are now upcastable to a "Sealed" supertrait (orange background), this is a recurring pattern (e.g., `std::io::IsTerminal`).

@rustbot labels +I-libs-api-nominated

I am nominating for libs API. Please take a look at that list and nominated for lang if it raises any immediate concerns.

**Comment by bjorn3** [comment]
`Error` should be 12 pointers (4 `Error` methods + 1 `Display` method + 1 `Debug` method + 2 * 3 header fields) big I believe and with upcasting forbidden it would be 9 pointers (4 `Error` methods + 1 `Display` method + 1 `Debug` method + 1 * 3 header fields). I don't think this is an issue.

> 21 traits are now upcastable to a "Sealed" supertrait (orange background), this is a recurring pattern (e.g., std::io::IsTerminal).

`std::sealed::Sealed` doesn't have any methods. As such it doesn't add to the vtable size but instead has it's vtable overlap with the start of the parent vtable AFAIK. For example `IsTerminal`'s vtable is size, align, drop_in_place, is_terminal, and upcasting to `Sealed` will reuse this exact same vtable, but ignore `is_terminal`.

**Comment by QuineDot** [comment]
[I happened to notice that the `trait_upcasting` feature gate can be bypassed by the `coerce_unsized` feature gate.](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e2ef10e7e09621ff1eb3b193528d34cb)
```rust
#![feature(coerce_unsized)]
// No feature(trait_upcasting)

fn upcast_me<'a, T>(x: &'a T) -> &'a dyn Any
where
    T: ?Sized,
    &'a T: CoerceUnsized<&'a dyn Any>,
{
    x as &dyn Any
}

//...
let _ = upcast_me::<dyn Trait>(&());
```

Not sure if it's a problem per se, but I thought I'd mention it in case the leakage was accidental.

**Comment by crlf0710** [comment]
This is known and not a problem (unless `coerced_unsized` feature got stablized before this, which is unlikely).

**Comment by nikomatsakis** [comment]
@crlf0710 I totally forgot about this, but now that [RFC #3324](https://github.com/rust-lang/rfcs/pull/3324) was merged, are we ready to stabilize?

**Comment by WaffleLapkin** [comment]
Although I would ***love*** to see this stabilized, I'm not sure if all concerns has been addressed, in particular:
- [ ] Exponential (?) blowup in size (raised @ https://github.com/rust-lang/rfcs/pull/3324#issuecomment-1305015681)
- [ ] Whatever we are fine with making creating `*const dyn Tr` with invalid metadata invalid, to support safe coercions
- [ ] Unresolved questions from this issue:
  - [ ] Should we make upcasting opt-in in some form to limit vtable size by default? The current inclination of the lang-team is "no", but it would be useful to gather data on how much supporting upcasting contributors to overall binary size.
  - [ ] Should we add an opt-out mechanism, and extend library stabilization checklist with "do we want to opt-out for now"?
- [ ] Whatever there are concerns with traits in the ecosystem outside of the standard library
 (raised @ https://github.com/rust-lang/rust/issues/65991#issuecomment-1369970129)
- [ ] While `trait A: B` is upcastable, should `trait A where Self: B` be? (raised @ https://github.com/rust-lang/rust/issues/65991#issuecomment-1294081137)

Some of those may have already been addressed/decided, but I could find where, if they even were. Anyway this is something the stabilization report / lang team will need to address.

**Comment by programmerjake** [comment]
well, afaict `*mut dyn Trait` with invalid metadata is unsound when combined with `arbitrary_self_types`, so imho that means invalid metadata should be documented as invalid anyway:
https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=fba1253fe76cfc4a573d6499aa9b56d4
the following compiles with no errors and calls using the metadata:
```rust
#![feature(arbitrary_self_types)]
pub trait Tr {
    fn f(self: *mut Self);
}

pub fn g(v: *mut dyn Tr) {
    Tr::f(v)
}
```

**Comment by nikomatsakis** [comment]
@WaffleLapkin thanks for the list! 

On the topic of size, I am still hoping for data about the size impact from supporting upcasting. All the machinery is on stable and has been for some time. I feel like there have been a lot of comments citing theoretical concerns but few with actual measurements -- but perhaps I've forgotten. 

Other folks on lang team may feel differently, but from my POV the size concerns are not a blocker for stabilization, more of a motivation tool to create better tooling or other mechanisms to reduce binary size for those contexts where it matters (e.g., it's already fairly common practice to strip some kinds of data in some contexts).

**Comment by crlf0710** [comment]
@WaffleLapkin thanks for the list! Just a note that the second bullet has been resolved via the fcp in https://github.com/rust-lang/rust/issues/101336 .

**Comment by WaffleLapkin** [comment]
> On the topic of size, I am still hoping for data about the size impact from supporting upcasting.

I think it should be possible to add a debug flag to the compiler, that prints `VTable` sizes with/without supporting upcasting and then run it on a set of crates somehow. I'll try to work in this direction.

---
<sup>@crlf0710 I've updated the list, thanks</sup>

**Comment by nikomatsakis** [comment]
@WaffleLapkin I like that idea! But I think we need a volunteer to do it.

**Comment by WaffleLapkin** [comment]
@nikomatsakis I have ideas how to make that work, this is near the top of my todo list =)

**Comment by nikomatsakis** [comment]
@WaffleLapkin great! My take is that I am willing to wait for people to gather data, but I'm not willing to block if nobody is signed up saying they will do it.

**Comment by nikomatsakis** [comment]
@rfcbot concern gathering data 

I opened this issue https://github.com/rust-lang/rust/issues/112355 to track @WaffleLapkin's work on gathering data.

**Comment by WaffleLapkin** [comment]
While working on #112355 I _think_ I found a bug. Given this program:
```rust
#![feature(rustc_attrs)]

#[rustc_dump_vtable]
pub trait What: Send + Sync {
    fn a(&self) {}
}

impl What for () {}

fn main() {
    (&() as &dyn What).a();
}
```
rustc outputs the following vtable layout:
```
 [
    MetadataDropInPlace,
    MetadataSize,
    MetadataAlign,
    TraitVPtr(<() as Sync>),
    Method(<() as What>::a),
]
```

Which is surprising, because `TraitVPtr(<() as Sync>)` is unnecessary — `Send` does not have any methods, so `Sync` vtable can be inlined...

**Comment by nikomatsakis** [comment]
Discussed in the rust-lang/lang meeting. To all those listening, @WaffleLapkin has prepared a flag that can gather data on the size of vtables, but nobody has tried it. We encourage everyone listening on this issue who has concerns to try this out on their project and post results in #112355 (instructions can be found in that issue). We are going to revisit this next week and if we don't see any data, or if the data we see doesn't concern us, we'll begin moving forward with FCP.

On a related note, I filed #113840 for the potential bug you found, @WaffleLapkin.

**Comment by WaffleLapkin** [comment]
While experimenting with vtables I also discovered #114007 (not _necessarily_ important for this, but still potentially interesting)

**Comment by nikomatsakis** [comment]
@rfcbot resolve gathering data

Based on what we see in #112355,  I feel comfortable moving forward now (and potentially doing follow up work later).

**Comment by pnkfelix** [comment]
Just wanted to quickly mention: I noticed this paper, "Tighten Rust's Belt: Shrinking Embedded Rust Binaries", recently that explicitly calls out vtable sizes as a concern for minimizing the size of Rust binaries: https://sing.stanford.edu/site/publications/rust-lctes22.pdf

**Comment by pnkfelix** [comment]
Temporarily removing T-compiler label to do a T-lang focused FCP.

@rustbot label: -T-compiler

**Comment by pnkfelix** [comment]
Following up on: https://github.com/rust-lang/rust/issues/65991#issuecomment-1640732308

After a few consecutive weeks discussing of the data presented in https://github.com/rust-lang/rust/issues/112355, T-lang believes this is ready to stabilize.

[2023-08-01 hackmd](https://hackmd.io/eEpxNWhQTauw9LzsrZqBkQ#%E2%80%9CTracking-issue-for-dyn-upcasting-coercion%E2%80%9D-rust65991)

[2023-08-08 hackmd](https://hackmd.io/VY0E31zbTMCRXx-BZRluxA#%E2%80%9CTracking-issue-for-dyn-upcasting-coercion%E2%80%9D-rust65991)

(The basic conclusion is that we think the cost of this will be limited enough that we can, if necessary, invest in an opt-out mechanism *at some future date*; we do not need to block stabilization today on the addition of a hypothetical opt-in mechanism, because we think migrating to an opt-in mechanism will be overly disruptive to Rust developers.)

@rfcbot fcp merge

**Comment by pnkfelix** [comment]
(readding T-compiler label now that I've started FCP merge.)

@rustbot label: +T-compiler

also, cc @rust-lang/compiler , if you have any concerns about this feature, make sure you get them reflected by a representative on T-lang.

**Comment by lcnr** [comment]
also, cc @rust-lang/types this feature needs pretty much the most involved builtin impls we have. `Unsize<dyn SuperTrait>: dyn Trait` is non-trivial , see e.g. https://github.com/rust-lang/rust/pull/114036. I still think this feature is very much desirable and this is a cost we are willing to pay.

**Comment by lcnr** [comment]
To add to the above point, I would like us[^1] to increase our requirements for stabilizing new features

#114036 dealt with an issue which is both quite subtle - the original approach in that PR added new ways to exploit #57893 - and the general issue of "projection bounds on trait objects have to be handled in some way", was afaict never mentioned anywhere during the stabilization process.

I have some ideas here, like requiring the stabilization report to again fully describe everything that is stabilized, referencing the implementation. I think starting the FCP without any summary of *what* is stabilized, or stating "implements RFC X", is bound to cause bigger issues going forward. But these ideas are my own and might not necessarily work out too well in practice. The types team will have an in-person meetup this year where I will try to talk about this and I am also interested in the opinions of others here via zulip or in-person.

[^1]: mostly t-types as that is where I am most involved, but also more generally t-compiler/t-lang etc

**Comment by crlf0710** [comment]
## Stabilization report

This fcp proposes the stablization of the `dyn` trait upcasting coercion feature, which is a long-wished-for feature [since 2013](https://github.com/rust-lang/rfcs/issues/2765), and tracked with the [Dyn upcasting coercion initiative](https://github.com/rust-lang/dyn-upcasting-coercion-initiative). The associated Reference PR in flight is 
https://github.com/rust-lang/reference/pull/1259.

We are stablizing a new kind of coercion, which is "upcasts" from a `dyn Trait` to its supertraits:

```rust
trait Foo {
    fn foo_method(&self);
}
trait Bar: Foo {
    fn bar_method(&self);
}

let x: &dyn Bar = /* ... */;
let y: &dyn Foo = x; // compiles
```

The key factor for these upcasts is that they require adjusting the vtable. The current implementation strategy is that the vtables for the `Bar` trait embed pointers to a `Foo` vtable within them:

```
+----------------------------+
| Bar vtable for some type T |
|----------------------------|
| Foo vtable                 | ----------> +------------------+
| foo_method                 | -----+      | Foo vtable for T |
| bar_method                 | --+  |      |------------------|
+----------------------------+   |  |      | foo_method       | ---+
                                 |  |      +------------------+    |
                                 |  |                              |
                                 |  +---> <T as Foo>::foo_method <-+
                                 v
                     <T as Bar>::bar_method
                 
(this diagram is only meant to convey the general idea of the vtable
 layout, and doesn't represent the exact offsets we would use etc;
 in fact, with the current implementation,
 the first supertrait is stored 'inline' and hence no load is required)
```

This way, given a `&dyn Bar` object, we convert its `Bar` vtable to the appropriate `Foo` vtable by loading the appropriate field.

For some examples of code patterns that this feature enables, see [Integration with `Any`: You can give any trait object `downcast` ability by using `Any` as its supertrait](https://github.com/rust-lang/dyn-upcasting-coercion-initiative/issues/2) and [Aligning type-erased closure bounds: You can use `&mut dyn Fn` as if it's a `&mut dyn FnMut`](https://github.com/rust-lang/dyn-upcasting-coercion-initiative/issues/3).

**Comment by lcnr** [comment]
Sorry for using this topic for the far more general topic of "I have issues with our stabilization process". My issues are not limited to this stabilization. They are not *mistakes* of whoever is pushing towards stabilization, but much rather issues with our existing process.

Using this stabilization report as an example of what I would ideally expect of features like this, based on my current - somewhat limited - knowledge about this feature:
- there seems to be a finished FCP on a not yet closed issue #101336. I would expect the decision made in that FCP to be referenced directly in the stabilization report with a short summary of its impact. Should this issue be closed before finishing the stabilization?
- afaik there has been significant discussion about the size of the vtables. T-lang decided that this size cost was acceptable: https://github.com/rust-lang/rust/issues/65991#issuecomment-1670127881. However, this decision and its rationale are pretty spread out and hard to follow. I would have liked a self-contained summary for that: linking to the collected data and summarizing the arguments used when making the decision
- afaict enabling this feature cause a breaking change for `Deref` impls from `Trait` to `SuperTrait`. There is a future compat lint for impls like this in #89460. I would like to see mention of this breakage and its impact (e.g. linking to crater run)
- "We are stablizing a new kind of coercion", an explanation of these coercion rules in detail. This should have ideally caught the issues fixed by #114036. This should have made it apparent that we currently don't support `dyn Trait + Send` to `dyn Send` upcasting, cc #114679.

Some of these additional requirements may require in-depth knowledge in many different areas, more than what can be expected from a single contributor. E.g. to accurately describe the coercion rules you probably need a @rust-lang/types member to participate in the stabilization report. However, such expertise is required anyways to evaluate the stabilization. It feels a lot better to me to potentially delay the stabilization because of this, rather than hoping that an expert notices stabilization and that - and how - it impacts their area of expertise.

**Comment by RalfJung** [comment]
Uh, is it expected that this [code using upcasting](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=91300effc70c6b43fc51ec02593a2851) works on nightly without any feature gates...? It actually builds on stable. :fearful:

**Comment by compiler-errors** [comment]
@RalfJung: That's not upcasting. That's unsizing `Box<dyn FileDescriptor>` to `dyn Any`.

**Comment by scottmcm** [comment]
@rfcbot concern ensure-we-discuss-lcnr's-thoughts-in-triage

**Comment by pnkfelix** [comment]
Discussed in T-lang meeting. We believe this is still S-waiting-on-team, its just waiting on T-types instead of T-lang.

@rustbot label: +I-types-nominated +T-types

**Comment by scottmcm** [comment]
This was discussed in triage, so I'll remove the concern
@rfcbot resolve ensure-we-discuss-lcnr's-thoughts-in-triage

**Comment by apiraino** [comment]
Were all concerns raised by @WaffleLapkin [in this comment](https://github.com/rust-lang/rust/issues/65991#issuecomment-1529933162) ticked/discussed? 

(apologies in advance if I missed something, I am slightly struggling to follow the technical details of the discussion)

**Comment by tmandry** [comment]
@apiraino I believe those should be addressed in some form in the stabilization report by the time it's published. I think T-types is now driving this to completion, but a number of other people have contributed to the draft (https://hackmd.io/QggP6SJVTa2jb7DjGq0vrQ?both).

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
@rfcbot concern types-team-sign-off

I am adding a concern that @rust-lang/types wants to be happy with the impl details (and especially @lcnr) -- @compiler-errors covered the details of type system interactions in rust-lang/rustc-dev-guide#1817 though.

**Comment by lcnr** [comment]
I am happy with the impl details. I had some other concerns in https://github.com/rust-lang/rust/issues/65991#issuecomment-1672782595.

While #101336 has still not been closed, everything seems now documented in https://hackmd.io/QggP6SJVTa2jb7DjGq0vrQ, so :+1: from my end.

**Comment by nikomatsakis** [comment]
@rfcbot resolve  types-team-sign-off

[see add'l discussion here](https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/2023-11-06.20planning.20meeting/near/400563434)

**Comment by GregoryConrad** [comment]
Not sure this has been mentioned yet, but seems related: #32220 should be possible now that Rust has upcasting support (at least in nightly)?

**Comment by nikomatsakis** [comment]
We discussed @WaffleLapkin's list of unresolved concerns in the rust-lang triage meeting and had this text which I will add to the stabilization report:


- [x] Exponential (?) blowup in size (raised @ https://github.com/rust-lang/rfcs/pull/3324#issuecomment-1305015681)
    - Conclusion: We put out a request for data in [#112355](https://github.com/rust-lang/rust/issues/112355). The data gathered did not suggest major impact (e.g., nbdd0121 found <0.2% overhead in the size of text sections). Furthermore, the vtables in practice have included upcast support for a long time and so this is not making anything *larger*.
- [x] Whatever we are fine with making creating `*const dyn Tr` with invalid metadata invalid, to support safe coercions (resolved by https://github.com/rust-lang/rust/issues/101336#issuecomment-1250356226)
- [x] Should we make upcasting opt-in in some form to limit vtable size by default? The current inclination of the lang-team is "no", but it would be useful to gather data on how much supporting upcasting contributors to overall binary size.
    - Answer: no. This would create complexity for all Rust users with very little practical benefit (size impact, as measured above, is very little).
- [x] Should we add an opt-out mechanism, and extend library stabilization checklist with "do we want to opt-out for now"?
    - Answer: we are not adding a mechanism but we can do it later. We wouldn't be able to apply it to the standard library traits but we did an analysis (in the stabilization report) and found that we would not want that anyway.
- [x] Whatever there are concerns with traits in the ecosystem outside of the standard library
 (raised @ https://github.com/rust-lang/rust/issues/65991#issuecomment-1369970129)
    - Answer: No, based on our data, we are not concetrned.
- [x] While `trait A: B` is upcastable, should `trait A where Self: B` be? (raised @ https://github.com/rust-lang/rust/issues/65991#issuecomment-1294081137)
    - Answer: Yes, and it is, because all other places in Rust treat `trait A: B` as equivalent to `trait A where Self: B`.

**Comment by tmandry** [comment]
@rfcbot reviewed

**Comment by NobodyXu** [comment]
If the upcasting feature isn't used, can it be optimized out by compiler?

**Comment by WaffleLapkin** [comment]
@NobodyXu if a type is not ever coerced to a trait object, then the compiler won't generate a vtable. Otherwise, if a type _is_ coerced, then the compiler will not be able to change the layout of the vtable based on it's usage.

**Comment by NobodyXu** [comment]
Thanks, it makes sense since it's hard to do that.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

The [proposal](https://github.com/rust-lang/rust/issues/65991#issuecomment-1670127881) to do this is in FCP.  Let's remove the nomination until and unless there is something more to do here.

**Comment by Mergulsevdova** [comment]
> This is a tracking issue for [RFC3324](https://github.com/rust-lang/rfcs/pull/3324). Corresponding [MCP](https://github.com/rust-lang/lang-team/issues/98) is here.
> 
> The feature gate for the issue is `#![feature(trait_upcasting)]`.
> 
> 
> 
> ### STATUS UPDATE
> 
> 
> 
> #### We are in the process of stabilizing.
> 
> 
> 
> ### About tracking issues
> 
> 
> 
> Tracking issues are used to record the overall progress of implementation.
> 
> They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
> 
> A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
> 
> Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.
> 
> 
> 
> ### Steps
> 
> 
> 
> - [x] Implement the experimental feature behind a feature gate.
> 
> - [x] Define the semantics explicitly.
> 
> - [x] Gathering design docs into initiative repo: https://github.com/rust-lang/dyn-upcasting-coercion-initiative
> 
> - [x] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide]) https://github.com/rust-lang/reference/pull/1259
> 
> - [x] Gather data about memory usage? https://github.com/rust-lang/rust/issues/112355
> 
> - [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide]) https://github.com/rust-lang/rust/pull/101718
> 
> 
> 
> ### Previous discussions
> 
> * RFC401: https://github.com/rust-lang/rfcs/pull/401
> 
> * RFC982: https://github.com/rust-lang/rfcs/pull/982
> 
> * RFC-issue2765: https://github.com/rust-lang/rfcs/issues/2765
> 
> * MCP: https://github.com/rust-lang/lang-team/issues/98 (Current lang-team MCP)
> 
> * Initiative: [Dyn upcasting coercion initiative](https://github.com/rust-lang/dyn-upcasting-coercion-initiative)
> 
> * RFC: https://github.com/rust-lang/rfcs/pull/3324
> 
> 
> 
> ### Unresolved Questions
> 
> * ~~Currently `CoerceUnsized` trait cannot express this case: a smart pointers that wrap a raw pointer and don't guarantee via a custom invariant that it is valid. Maybe a separate `CoerceUnsizedUnsafe` trait is needed. (see https://github.com/rust-lang/rust/pull/88010#issuecomment-898893209).~~
> 
> * Should we make upcasting opt-in in some form to limit vtable size by default? The current inclination of the lang-team is "no", but it would be useful to gather data on how much supporting upcasting contributors to overall binary size.
> 
> * ~~Before stabilizing it we should check that libs-api is ok with upcasting for all dyn-allowed traits in the library, since those we can't change.~~ (addressed by https://github.com/rust-lang/rust/issues/65991#issuecomment-1369970129)
> 
> * Should we add an *opt-out* mechanism, and extend library stabilization checklist with "do we want to opt-out for now"?
> 
> 
> 
> ### Implementation history
> 
> * Initial attempt: #60900
> 
> * Second attempt:
> 
>   * [X] Preparation refactoring: #86291
> 
>   * [X] Unify vtable building: #86475
> 
>   * [x] Update vtable format: #86461
> 
>   * [x] Feature gate implementation: #86264
> 
>   * [x] Pointer conversion adjustments: #87515
> 
>   * [x] Support multiple occurrence of a target trait with different substsdefs: #88135
> 
>   * [x] ~~Unsafety checking for raw dyn pointers: #88010 (not prefered approach, likely will need to do nothing here, but need a design meeting discussion with lang-team)~~ Solution: https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/meeting.202022-09-02.20planning/near/296826711, https://github.com/rust-lang/rust/issues/101336
> 
>   * [x] Lint for easing migration: https://github.com/rust-lang/rust/issues/89460
> 
>   * [x] Reference changes https://github.com/rust-lang/reference/pull/1259 
> 
>   * [x] Mark this feature no longer incomplete and stablize: https://github.com/rust-lang/rust/pull/101718
> 
>   * [x] Make trait upcasting work with supertraits with fewer associated types: #114036

-

**Comment by Ekleog** [comment]
I just tried using this feature to upcast to `Any + Send + Sync`, and it seems to me like this is currently not allowed:
```rust
trait DynSized: 'static + Any + Send + Sync + DeepSizeOf {}
impl<T: 'static + Any + Send + Sync + DeepSizeOf> DynSized for T {}

fn main() {
    let ptr = Arc::new(()) as Arc<dyn DynSized>;
    let ptr_any = ptr as Arc<dyn Any + Send + Sync>;
    // my goal:
    let res = Arc::<dyn Any + Send + Sync>::downcast::<()>(ptr);
}
```
[(playground link)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d8992f483c4ac73610acffe8dc0e814b)

With references, the situation does not seem better than with `Arc`: this does not compile either
```rust
fn main() {
    let ptr = &() as &'static dyn DynSized;
    let ptr_any = ptr as &'static (dyn Any + Send + Sync);
}
```
[(playground link)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c59ca2feb371772118f0c4f4b4599a00)

Is this expected? I can't find any reference to this limitation, neither explicitly in the RFC text nor in this tracking issue. The RFC text does mention not supporting upcasting to multiple traits, but I'd have expected lifetimes as well as auto traits to not be included in the limitation.

If it is expected that this feature is stabilized without this support, does anyone know where I should watch, to know of the news on that front?

**Comment by GregoryConrad** [comment]
@Ekleog you can probably do a workaround like:

```rust
trait AnySendSync: Any + Send + Sync {}
impl<T: Any + Send + Sync> AnySendSync for T{}
```

Then upcast to AnySendSync instead of Any + Send + Sync. On my phone though so can’t verify if this actually works

**Comment by Ekleog** [comment]
Unfortunately, it looks like [I still cannot downcast, because I need to upcast `Arc<dyn AnySendSync>` to `Arc<dyn Any + Send + Sync>` in order to be able to downcast, and [this still does not compile](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=68545bd650a798aa4fac79a5b97f946f) :/

**Comment by CAD97** [comment]
It's certainly not ideal due to requiring `unsafe`, but a `downcast_ref` to check followed by a `Arc::from_raw(arc.as_raw().cast())` would at least be be a sound workaround.

**Comment by yshui** [comment]
@Ekleog how about this? this has an extra `T: Sized` restriction though, don't know if it can be better.

```rust
use std::any::Any;
use std::sync::Arc;

trait DeepSizeOf {} // from deepsize
impl DeepSizeOf for () {}

trait AnySendSync: Any + Send + Sync {
    fn explode(self: Arc<Self>) -> Arc<dyn Any + Send + Sync>;
}
impl<T: Any + Send + Sync + Sized> AnySendSync for T {
    fn explode(self: Arc<T>) -> Arc<dyn Any + Send + Sync> {
        self
    }
}
trait DynSized: 'static + AnySendSync + DeepSizeOf {}
impl<T: 'static + AnySendSync + DeepSizeOf> DynSized for T {}

fn main() {
    let ptr = Arc::new(()) as Arc<dyn DynSized>;
    let ptr_anysendsync = ptr as Arc<dyn AnySendSync>; // works fine
    let ptr_any = ptr_anysendsync.explode();
    let res = ptr_any.downcast::<()>();
}
```

**Comment by RalfJung** [comment]
Closed issues are not a good place for discussion, please open a new issue or discuss on Zulip or the forums.

> upcasting to a group of traits isn't possible right now

Indeed that's the core here, you can only upcast to a single trait. Could be interesting to discuss extending this to "one trait plus any number of marker traits" but that's an extension which needs a new issue. :)

**Comment by dhardy** [comment]
Investigating @Ekleog's problem, I see two issues.

### Dyn object cast with marker traits

```rust
error[E0605]: non-primitive cast: `Arc<dyn DynSized>` as `Arc<(dyn Any + Send + Sync + 'static)>`
  --> src/main.rs:12:19
   |
12 |     let ptr_any = ptr as Arc<dyn Any + Send + Sync>;
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
```
Rust does not (at this time) support [multi-trait objects](https://github.com/rust-lang/rfcs/issues/2035), with an exception for [auto traits](https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits).

The RFC notes that upcast to arbitrary multi-trait objects like `dyn (Trait1 + Trait2)` is not supported, but does not mention upcast to `dyn (Trait1 + A)` for auto-trait(s) `A`. **Can this RFC be extended for upcast to a trait + auto-traits?**

### Downcast

Methods [`Any::is`](https://doc.rust-lang.org/stable/std/any/trait.Any.html#method.is) and [`Box::downcast`](https://doc.rust-lang.org/stable/std/any/trait.Any.html#method.downcast) (also `Arc::downcast`, etc.) are not defined on the trait but on various objects.

To make them flexible enough I believe we'd need a new form of generics (and that the modification below would be a breaking change), so it's not an easy problem to solve:
```rust
// NOTE: "dyn generics"
impl<dyn N: Any, A: Allocator> Box<T, N> {
    pub fn downcast<T: Any>(self) -> Result<Box<T, A>, Self> {
        // ...
    }
}
```

**Comment by Ekleog** [comment]
Let's continue discussion on my upcasting issue on [this separate issue opened by @compiler-errors](https://github.com/rust-lang/rust/issues/119335) :) Incidentally, it is currently being solved in [this PR also by @compiler-errors](https://github.com/rust-lang/rust/pull/119338) (lots of thanks :heart:)

@dhardy As for the downcasting you're mentioning, I'm not sure what you're thinking of? But if/when you open a separate issue to discuss it, please tag me!

**Comment by dhardy** [comment]
@Ekleog regarding `downcast`, see https://internals.rust-lang.org/t/pre-rfc-make-downcast-user-friendly/20092

**Comment by steffahn** [comment]
I’ve come across some soundness issues with upcasting today. Readers of this thread might be interested ^^

* https://github.com/rust-lang/rust/issues/120222

**Comment by WaffleLapkin** [comment]
The stabilization is currently blocked on the fix of the unsoundness bug tracked in #120222. The fix is in #120248, but needs a bit more work. I've been occupied by the never type stuff which is urgent because some of the changes are edition based. I hope I'll have time soon to finish #120248.

This is _also_ blocked on rust reference PR, which someone needs to take over:  https://github.com/rust-lang/reference/pull/1259#issuecomment-1907299080.

**Comment by Kish29** [comment]
It looks like the blocking bug #120222  have been fixed out, and what's next for this stabilization?

**Comment by WaffleLapkin** [comment]
@Kish29 the next steps are changing the reference (cc https://github.com/rust-lang/reference/pull/1259) and doing a new stabilization PR :)

**Comment by noclue** [comment]
I am newbie baffled by the absence of this basic syntax. I got this workaround if anyone is interested. It works all on stable Rust

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0d53066544591aadc71e22bc17f0bb2a
